tket/src/WeightSubgrMono/EndToEndWrappers/MainSolver.cpp

```diff
@@ -0,0 +1,421 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "WeightSubgrMono/EndToEndWrappers/MainSolver.hpp"
+
+#include <algorithm>
+#include <chrono>
+#include <numeric>
+
+#include "Utils/Assert.hpp"
+#include "WeightSubgrMono/Common/GeneralUtils.hpp"
+#include "WeightSubgrMono/EndToEndWrappers/PreSearchComponents.hpp"
+#include "WeightSubgrMono/EndToEndWrappers/SearchComponents.hpp"
+#include "WeightSubgrMono/GraphTheoretic/DomainInitialiser.hpp"
+#include "WeightSubgrMono/WeightPruning/WeightChecker.hpp"
+
+namespace tket {
+namespace WeightedSubgraphMonomorphism {
+
+typedef std::chrono::steady_clock Clock;
+
+MainSolver::MainSolver(
+    const GraphEdgeWeights& pattern_edges, const GraphEdgeWeights& target_edges,
+    const MainSolverParameters& parameters)
+    : m_pattern_vertex_relabelling(pattern_edges),
+      m_target_vertex_relabelling(target_edges),
+      m_pattern_neighbours_data(
+          m_pattern_vertex_relabelling.new_edges_and_weights),
+      m_target_neighbours_data(
+          m_target_vertex_relabelling.new_edges_and_weights) {
+  const auto num_p_vertices =
+      m_pattern_neighbours_data.get_number_of_nonisolated_vertices();
+  if (num_p_vertices == 0) {
+    m_solution_data.trivial_weight_lower_bound = 0;
+    m_solution_data.trivial_weight_initial_upper_bound = 0;
+    m_solution_data.finished = true;
+    return;
+  }
+  const auto num_t_vertices =
+      m_target_neighbours_data.get_number_of_nonisolated_vertices();
+  {
+    const auto number_of_possible_t_edges =
+        (num_t_vertices * (num_t_vertices - 1)) / 2;
+    m_solution_data.target_is_complete =
+        number_of_possible_t_edges ==
+        m_target_neighbours_data.get_number_of_edges();
+  }
+
+  // Start off assuming that it's impossible. So L = +inf, U = 0 make
+  // sense mathematically (infimum over empty set is +infinity, etc. etc.)
+  m_solution_data.trivial_weight_initial_upper_bound = 0;
+  set_maximum(m_solution_data.trivial_weight_lower_bound);
+
+  if (m_pattern_neighbours_data.get_number_of_edges() >
+          m_target_neighbours_data.get_number_of_edges() ||
+      num_p_vertices > num_t_vertices) {
+    m_solution_data.finished = true;
+    return;
+  }
+
+  const auto init_start = Clock::now();
+
+  m_pre_search_components_ptr = std::make_unique<PreSearchComponents>(
+      m_pattern_neighbours_data, m_target_neighbours_data);
+  TKET_ASSERT(m_pre_search_components_ptr);
```

Comment:
OK, this is a different issue; I am worried that `make_unique` might return a null ptr rather than throwing `std::bad_alloc` if it cannot create the object.

Probably (hopefully) this is forbidden by the standard, but the docs are NOT 100% clear on this point.